### PR TITLE
Fight! style overhaul

### PR DESCRIPTION
**Pull request by Nivroc**
##Changes
This PR overhauls the look of Fight! dialog window. It optimises space and structures information, allowing for further Fight! additions like wounds, stance, tax, 1vX penatly, size Lock penalty ect..
Most notable changes are:
- combatants details(positon, engagement ect.) are now structured as a table
- combatants portraits are moved to overview
- hiding combatants for GM means overview shows only details, but not portrait or roll buttons
- Steel attribute roll is included to overview panel
- most HTML elements are now treated as flex to promote reactive design and look nicer on small resolutions/window sizes
- overview is able to show larger portraits for both GM and players

Possible TODOs: several fight tabs with Eye of the storm! feature, "sides" tag on actors, auto volley resolution for sides

## Extra Info

Extra information for the fix is also quite helpful. Try to answer any relevant questions in the comment.
- [No] Did this PR have to change `template.yml`?
- [- ] If so, were there any steps taken to protect existing user data? A migration task?
- [X] Did you follow the [contributing guidelines](https://github.com/StasTserk/foundry-burningwheel/blob/master/CONTRIBUTING.md)?
- [No] Is this PR limited to fixing just one issue? Although there are several changes they all are related only to Fight! dialog restyle

### CODE REVIEW

**Comment by StasTserk** [comment]
Going to delve into the code next, but from a cursory perusal of the PR, a few issues of issues with the UI stand out...
- [ ] Extra long names aren't handled well in the bottom grid
image

- [ ] The border around portraits is causing problems with non-square portraits.
image

- [ ] As per our discussion, adding the character portrait for npcs hidden from the GM would be rather nice. Only have it shift the first 3 rows, keep the weapon row separate.
image
to
image

File: module/dialogs/FightDialog.ts [hunk 1]
```diff
@@ -111,19 +111,19 @@ export class FightDialog extends ExtendedTestDialog<FightDialogData> {
...
+        ["Speed", "Power", "Agility", "Skill", "Steel"].forEach((attr: string) => {
+            html.find('button[data-action="roll'+attr+'"]')
+                .on('click', (e: JQuery.ClickEvent) => { this._handleRoll(e, attr.toLowerCase() as constants.FightAttr); });
         });
+        const openSheet = (e: JQuery.ClickEvent) => {
+            const id = e.currentTarget.attributes.getNamedItem("data-actor-id").nodeValue || "";
+            game.actors.find(a => a._id === id).sheet.render(true);
+        };
+        html.find('img[data-action="openSheet"]').on('click', openSheet);
+        html.find('div[data-action="openSheet"]').on('click', openSheet);
```
Comment (StasTserk) [inline module/dialogs/FightDialog.ts:123]:
```suggestion
        html.find('div[data-action="openSheet", img[data-action="openSheet"]').on('click', (e: JQuery.ClickEvent) => {
            const id = e.currentTarget.attributes.getNamedItem("data-actor-id").nodeValue || "";
            game.actors.find(a => a._id === id).sheet.render(true);
        });
```
You can condense the selector to just one line here

File: module/dialogs/FightDialog.ts [hunk 2]
```diff
@@ -111,19 +111,19 @@ export class FightDialog extends ExtendedTestDialog<FightDialogData> {
...
     }
     
-    private _handleRoll(e: JQuery.ClickEvent, type: "speed" | "agility" | "power" | "skill") {
+    private _handleRoll(e: JQuery.ClickEvent, type: constants.FightAttr) {
```
Comment (StasTserk) [inline module/dialogs/FightDialog.ts:126]: I think it's probably fine to pull `FightAttr` from `constants.ts` into this file. I don't foresee anyone outside of the fight dialog needing to reference it other than the one use in `FightRoll.ts`. Open to thoughts on this though

File: module/dialogs/FightDialog.ts [hunk 3]
```diff
@@ -253,7 +253,7 @@ export interface FightDialogData {
     showV3: boolean;
 }
 
-interface ParticipantEntry {
+export interface ParticipantEntry {
```
Comment (StasTserk) [inline module/dialogs/FightDialog.ts:256]: as per the helper ealier comment, I don't believe this is necessary.

File: module/constants.ts [hunk 4]
```diff
@@ -46,6 +46,8 @@ export type SkillTypeString =
     "musical" | "peasant" | "physical" | "schoolofthought" |
     "seafaring" | "special" | "social" | "sorcerous" | "training";
 
+export type FightAttr = "speed" | "agility" | "power" | "skill" | "steel";
```
Comment (StasTserk) [inline module/constants.ts:49]: I don't foresee anyone other than the fight dialog and `FightRoll.ts` needing this. I think it's fine to just pull into FightDialog.ts

File: templates/dialogs/fight.hbs [hunk 5]
```diff
@@ -21,17 +21,55 @@
...
+                    <button class="control-button" data-index="{{i}}" data-action="rollSpeed">Speed</button>
+                    <button class="control-button" data-index="{{i}}" data-action="rollPower">Power</button>
+                    <button class="control-button" data-index="{{i}}" data-action="rollAgility">Agility</button>
+                    <button class="control-button" data-index="{{i}}" data-action="rollSkill">Skill</button>
+                    <button class="control-button" data-index="{{i}}" data-action="rollSteel">Steel</button>
+                </div>
+                {{else}}
+                    {{#if ../gmView}}
+                        <div class="details clickable" data-action="openSheet" title="{{p.name}}" data-actor-id="{{p.id}}">
+                        <table>
```
Comment (StasTserk) [inline templates/dialogs/fight.hbs:42]: Hard 👎 on this design. You might note that there are exactly no tables being used in burning wheel. If you want to organize things into a grid, use a div with `display: grid`.

File: styles/dialog/fight.scss [hunk 6]
```diff
@@ -37,17 +92,42 @@
...
+                    align-items: center;
+                    overflow: hidden;
+                    max-height: 130px;
+                    .control-button {
+                        overflow: hidden;
+                        width: 49%;
+                    }
+                }
             }
+           // .control-button {
```
Comment (StasTserk) [inline styles/dialog/fight.scss:115]: this can be deleted

File: styles/dialog/fight.scss [hunk 7]
```diff
@@ -7,9 +7,9 @@
     }
 
     .control-grid {
-        display: grid;
-        grid-template-columns: 1fr 1fr;
-        column-gap: 10px;
+        display: flex;
+        align-items: center;
+        justify-content: center;
```
Comment (StasTserk) [inline styles/dialog/fight.scss:12]: I'm not sure why this change was necessary. The only thing that is visually different is the fact there is no `column-gap` anymore.

File: styles/dialog/fight.scss [hunk 8]
```diff
@@ -23,9 +23,64 @@
         display: grid;
         width: 100%;
         height: min-content;
+        overflow: hidden;
```
Comment (StasTserk) [inline styles/dialog/fight.scss:26]: Ditto with this change. This hides any part of the grid that displays the fight cards if it happens to overflow its' maximum height. It doesn't have a maximum height though.

File: styles/dialog/fight.scss [hunk 9]
```diff
@@ -23,9 +23,64 @@
...
+
+        .details {
+            display: flex;
+            align-items: center;
+            justify-content: center;
+            flex-grow: 1;
+            .reflexes {
+                padding-right: 15px;
+            }
+            td {
```
Comment (StasTserk) [inline styles/dialog/fight.scss:58]: I believe in the case of the Fight dialog, this is misuse of the rather narrow use of the `table` tag and all the related bits, so 👎 on including this in its' current state.

**Comment by Nivroc** [comment]
1) border is removed
2) I can add padding, but participant names will no longer be centered. Took a bit of work to center them with absolute and relative positions of 'remove'. It seems it's either long names colliding with 'remove' or slightly uncentered names. if you know a way to solve this let me know. If not please select the preferred choice. Personally I'd go with colliding names.
3) The idea of this PR is to prepare Fight! dialog for further addition of useful details. To name a few: wounds, sorc tax, 1vX penalties, terrain advantages, stance, knock-down. When adding those I'll structure the details table to contain as many columns as needed(flexbox). Including portrait takes up space from those details. As a GM i'd personally prefer as much info as I can get in the most compact way. Portraits are still available on "non-detailed" view, but when handling a complex fight I wouldn't want anything taking up space. 
I will probably move weapon name to be one row of otherwise multi-rowed details table in the next PR.

**Review by StasTserk — CHANGES_REQUESTED** [review]
I like the general look of what you're trying to accomplish, I think you have the general approach correct, but due to your use of tables in the html layout, this isn't something I'm willing to approve as is.

Couple of other general comments:
 - I see a few places where you've added `overflow: hidden` to the code. I'm not convinced that all of them belong there, especially on elements without a fixed height.
  - Though I'm not best at this either, try to avoid too many hard pixel values when it comes to text and inputs. Using a unit like `em` allows the code to better handle things like accessibility features that resize text to be more readable.
  - You've included a number of inline styles for elements. I think the case for doing this instead of something in a css class has to be really good, and I haven't run into it with this project yet.

**Comment by StasTserk** [comment]
I want to probably delve into the things going on in the css a bit deeper, but I don't think it makes sense to do that as long as there is layout work still to do to remove tables.

**Comment by StasTserk** [comment]
As replying to comment 2, you can have a look at how range and cover does each team's title. By default the names are super short but you can see what happens when you manually edit the name.

image

File: styles/dialog/fight.scss (hunk 7 above)
Comment (Nivroc) [reply styles/dialog/fight.scss:12]: Also it forms sort of a control-panel to which it's possible to add other buttons and the space will be divided evenly among them without having to every change css

File: module/dialogs/FightDialog.ts (hunk 2 above)
Comment (Nivroc) [reply module/dialogs/FightDialog.ts:126]: May be better to keep in the general list as if any further refactoring occurs(fight gets split into multiple files for whatever reason) it might get forgotten. Don't mind either way though - very minor thing

Comment (StasTserk) [reply module/dialogs/FightDialog.ts:126]: I'll leave this entirely up to you, but I think it's better to keep only broadly used constants in `constants.ts` so it doesn't become overwhelmed with bits from every single file that wants to define something. I would expect if fight gets split into multiple files somehow, there would still be some logical place to keep constants that are mostly related to the fight dialog itself.

File: module/dialogs/FightDialog.ts (hunk 1 above)
Comment (Nivroc) [reply module/dialogs/FightDialog.ts]: @StasTserk for some truly diabolical reason suggested change alone makes Fight! dialog have full screen width on open...

Comment (StasTserk) [reply module/dialogs/FightDialog.ts:125]: Missing left square bracket after the first part of the search.
image

File: module/dialogs/FightDialog.ts [hunk 10]
```diff
@@ -111,19 +113,17 @@ export class FightDialog extends ExtendedTestDialog<FightDialogData> {
...
-
-        html.find('button[data-action="rollSpeed"]').on('click', (e: JQuery.ClickEvent) => { this._handleRoll(e, "speed"); });
-        html.find('button[data-action="rollPower"]').on('click', (e: JQuery.ClickEvent) => { this._handleRoll(e, "power"); });
-        html.find('button[data-action="rollAgility"]').on('click', (e: JQuery.ClickEvent) => { this._handleRoll(e, "agility"); });
-        html.find('button[data-action="rollSkill"]').on('click', (e: JQuery.ClickEvent) => { this._handleRoll(e, "skill"); });
-
-        html.find('img[data-action="openSheet"]').on('click', (e: JQuery.ClickEvent) => {
-            const id = e.target.dataset.actorId || "";
+        ["Speed", "Power", "Agility", "Skill", "Steel"].forEach((attr: string) => {
+            html.find('button[data-action="roll'+attr+'"]')
```
Comment (StasTserk) [inline module/dialogs/FightDialog.ts:117]:
```suggestion
            html.find(`button[data-action="roll${attr}"]`)
```
Prefer string interpolation here

File: styles/dialog/fight.scss [hunk 11]
```diff
@@ -19,13 +19,71 @@
         }
     }
 
+    .details {
+        display: grid;
+        grid-template-columns: repeat(4, 1fr);
```
[the remark this answers is not in the document]
Comment (StasTserk) [reply styles/dialog/fight.scss:24]: This should also be a different style depending on if this is displaying the GM info fighter information at the top of the dialog as opposed to the player view below which has wider inputs that need to be filled out., splitting styles here is probably a good idea.

File: styles/dialog/fight.scss [hunk 12]
```diff
@@ -37,17 +95,38 @@
...
+                //justify-content: center;
+                //align-items: center;
+                .fight-rolls {
+                    flex-grow: 1;
+                    display: flex;
+                    flex-wrap: wrap;
+                    flex-direction: column;
+                    justify-content: center;
+                    align-items: center;
+                    overflow: hidden;
```
Comment (StasTserk) [inline styles/dialog/fight.scss:110]: Fairly certain this `overflow` tag is unnecessary.

File: styles/dialog/fight.scss [hunk 13]
```diff
@@ -101,37 +181,38 @@
...
-                select {
-                    width: 100%;
+                .actor-name {
+                    display: flex;
+                    justify-content: center;
+                    align-items: center;
+                    font-size: 1.125em;
+                    font-weight: bold;
+                    flex-grow: 1;
+                    padding-right: 1.5em;
```
Comment (StasTserk) [inline styles/dialog/fight.scss:202]: Not sure about this, especially when it comes to the player only view.
image

File: styles/dialog/fight.scss [hunk 14]
```diff
@@ -101,37 +181,38 @@
...
+                    width: calc(100% - 1.5em);
```
Comment (StasTserk) [inline styles/dialog/fight.scss:203]: can be deleted. `flex-grow` is meant to control the width of an element in a flex box.

File: styles/dialog/fight.scss [hunk 15]
```diff
@@ -101,37 +181,38 @@
...
-            }
-            .details {
-                width: calc(95% - 80px);
-                margin-left: 2%;
-                select {
-                    width: 100%;
+                .actor-name {
+                    display: flex;
+                    justify-content: center;
+                    align-items: center;
```
Comment (StasTserk) [inline styles/dialog/fight.scss:198]:
```suggestion
                    text-align: center;
```

File: templates/dialogs/fight.hbs [hunk 16]
```diff
@@ -74,32 +104,38 @@
...
-                    Reflexes: {{p.reflexes}}
-                </div>
-                <div>
-                    Engagement Bonus
-                    <input data-index="{{i}}" name="engagementBonus" type="number" class="exponent no-arrows" value="{{p.engagementBonus}}">
+                <div class="details-key">Reflexes</div>
+                <div class="details-value">{{p.reflexes}}</div>
+                <div class="details-key">Engagement</div>
+                <div class="withLabels details-value">
+                    <div class="exponent-meaning">D</div>
```
Comment (StasTserk) [inline templates/dialogs/fight.hbs:119]: `exponent-meaning` doesn't seem to appear anywhere, can be deleted.

File: templates/dialogs/fight.hbs [hunk 17]
```diff
@@ -74,32 +104,38 @@
...
-                <div>
-                    Reflexes: {{p.reflexes}}
-                </div>
-                <div>
-                    Engagement Bonus
-                    <input data-index="{{i}}" name="engagementBonus" type="number" class="exponent no-arrows" value="{{p.engagementBonus}}">
+                <div class="details-key">Reflexes</div>
+                <div class="details-value">{{p.reflexes}}</div>
+                <div class="details-key">Engagement</div>
+                <div class="withLabels details-value">
```
Comment (StasTserk) [inline templates/dialogs/fight.hbs:118]: `withLabels` doesn't seem to appear in the css at all, can be deleted.

File: templates/dialogs/fight.hbs (hunk 16 above)
Comment (StasTserk) [reply templates/dialogs/fight.hbs:119]: moreover, it's not clear to me why these appear in reverse order other than `flex-direction: row-reverse` is used in the `details-value` css class.

File: styles/dialog/fight.scss [hunk 18]
```diff
@@ -19,13 +19,71 @@
...
+            display: flex;
+            align-items: center;
+            background: rgba(#000, .03);
+            border: 2px groove #aaaaaa;
+            padding: 1px;
+        }
+        .details-value{
+            display: flex;
+            align-items: center;
+            flex-direction: row-reverse;
```
Comment (StasTserk) [inline styles/dialog/fight.scss:39]: Why reverse?

File: styles/dialog/fight.scss [hunk 19]
```diff
@@ -19,13 +19,71 @@
...
+        display: grid;
+        align-items: center;
+        padding: 5px 5px 0 5px;
+        //Weapon word has static width
+        grid-template-columns: [line1] 50px [line2] auto [col2-start];
+        justify-content: stretch;
+        .weapon-picker {
+            justify-self: end;
+            grid-column-start: 2;
+        }
```
Comment (StasTserk) [inline styles/dialog/fight.scss:58]: Perhaps something simpler
```suggestion
        display: grid;
        width: 100%;
        padding: 5px 5px 0 5px;
        //Weapon word has static width
        grid-template-columns: 50px 1fr;
        .weapon-picker {
            text-align: end;
        }
```

**Review by StasTserk — CHANGES_REQUESTED** [review]
Good work on getting rid of all the table stuff, I think there's more to be said about the css, but let's leave it here for now.

As stated in the actual review comments, I think the `details` css class shouldn't be the same for the volley and fighter grids. There is some commonality between them that should be extracted but their content is different enough to where matching styles doesn't quite work, I don't think.

On an aesthetic note, I'm not a huge fan of the use of borders on the details grid. Perhaps something a bit more light on borders can be used instead?

An example:
image

File: styles/dialog/fight.scss (hunk 11 above)
[the remark this answers is not in the document]
Comment (Nivroc) [reply styles/dialog/fight.scss:24]: given the logic of this display and the justifies/alignments described below, it's very responsive as of now. all keys have width of teh widest key and other elements align and stretch. There's no need to excpliticly describe I think.

File: styles/dialog/fight.scss (hunk 18 above)
Comment (Nivroc) [reply styles/dialog/fight.scss:39]: So that items "stick" to the right and not left by default.

File: styles/dialog/fight.scss (hunk 19 above)
Comment (Nivroc) [reply styles/dialog/fight.scss:58]: template-columns and width are good,
but text-align doesn't handle everything well

File: styles/dialog/fight.scss [hunk 20]
```diff
@@ -37,17 +95,38 @@
...
-                width: 49%;
-                height: min-content;
+            .portrait-actions {
+                flex-grow: 1;
+                display: flex;
+                //justify-content: center;
+                //align-items: center;
+                .fight-rolls {
+                    flex-grow: 1;
+                    display: flex;
```
[the remark this answers is not in the document]
Comment (Nivroc) [reply styles/dialog/fight.scss:105]: As I'm specifying more flex options here I think it would be clearer to have them all in one place so it's much easier to read and leverage. flex-column is just an alias for some of those properties.

File: styles/dialog/fight.scss (hunk 15 above)
Comment (Nivroc) [reply styles/dialog/fight.scss:198]: image

File: styles/dialog/fight.scss (hunk 13 above)
Comment (Nivroc) [reply styles/dialog/fight.scss:202]: It's the solution for the issue you've had before. Text is now always to center and not colliding with remove icon

File: styles/dialog/fight.scss (hunk 14 above)
Comment (Nivroc) [reply styles/dialog/fight.scss:203]: image

File: templates/dialogs/fight.hbs [hunk 21]
```diff
@@ -21,17 +21,47 @@
...
+                    <div class="fight-rolls">
+                        <button class="control-button" data-index="{{i}}" data-action="rollSpeed">Speed</button>
+                        <button class="control-button" data-index="{{i}}" data-action="rollPower">Power</button>
+                        <button class="control-button" data-index="{{i}}" data-action="rollAgility">Agility</button>
+                        <button class="control-button" data-index="{{i}}" data-action="rollSkill">Skill</button>
+                        <button class="control-button" data-index="{{i}}" data-action="rollSteel">Steel</button>
+                    </div>
+                {{else}}
+                    {{#if ../gmView}}
+                        <div class="gm-overview">
```
[the remark this answers is not in the document]
Comment (Nivroc) [reply templates/dialogs/fight.hbs:41]: While intendation was improved, just keeping width:100% on weapon block would not provide the same view.

**Comment by Nivroc** [comment]
Thought a lot about details split, but imho it's not the best time. My reasoning is:
Currently I see no need to distinguish the looks of the 2. Keeping only 1 CSS class without copy-pasting it 3 times to 3 separate parts of a huge CSS file just to get rid of a couple of options(which don't change the look) is doing more harm, than good.
If the need to change looks somehow arises via the property of css the needed option can be added below.
If there's a need for a major change of the looks the class can be split indeed.
For now I think the split will only mess up the already pretty long css file and make further development harder.

File: styles/dialog/fight.scss (hunk 15 above)
Comment (StasTserk) [reply styles/dialog/fight.scss]: I'm not seeing that at all.

image
image

**Review by StasTserk — APPROVED** [review]

**Comment by StasTserk** [comment]
:+1:

Good work on this, I'm going to make some adjustments to personal taste in a bit, but merging this for now.
